pkg/vmssrollout/upgrade.go

```diff
@@ -0,0 +1,324 @@
+package vmssrollout
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+)
+
+type ImageReference struct {
+	ID      string
+	SKU     string
+	Version string
+}
+
+type VMSSUpgrader struct {
+	SubscriptionID string
+	ResourceGroup  string
+	// Name of the VMSS to upgrade.
+	Name string
+
+	// provides an interface for configuring
+	// waiting for the readiness of a VM.
+	Plugin api.Upgrade
+
+	// upgrade parameters
+	DesiredReplicas int64
+	ImageRef        *ImageReference
+	Drain           bool
+	InPlace         bool
+	MaxSurge        int64
+	MaxUnavailable  int64
+
+	ssc compute.VirtualMachineScaleSetsClient
+	vmc compute.VirtualMachineScaleSetVMsClient
+}
+
+func (u *VMSSUpgrader) clientSetup() error {
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return err
+	}
+
+	ssc := compute.NewVirtualMachineScaleSetsClient(u.SubscriptionID)
+	ssc.Authorizer = authorizer
+	u.ssc = ssc
+
+	vmc := compute.NewVirtualMachineScaleSetVMsClient(u.SubscriptionID)
+	vmc.Authorizer = authorizer
+	u.vmc = vmc
+	return nil
+}
+
+func (u *VMSSUpgrader) Upgrade() error {
+	if err := u.clientSetup(); err != nil {
+		return err
+	}
+
+	vmss, err := u.ssc.Get(context.Background(), u.ResourceGroup, u.Name)
+	if err != nil {
+		return fmt.Errorf("cannot get VMSS %q in resource group %q: %v", u.Name, u.ResourceGroup, err)
+	}
+
+	if vmss, err = u.updateImageVMSS(vmss); err != nil {
+		return err
+	}
+
+	log.Printf("VMSS %q current capacity at %d", u.Name, *vmss.Sku.Capacity)
+
+	vmsToUpgrade, err := u.upgradable()
+	if err != nil {
+		return fmt.Errorf("cannot list upgradable VMs in VMSS %q: %v", u.Name, err)
+	}
+
+	for _, vmToUpgrade := range vmsToUpgrade {
+		// if possible, scale up first
+		canScaleUp, err := u.canScaleUp()
+		if err != nil {
+			return err
+		}
+		if canScaleUp {
+			log.Printf("Scale up VMSS %q", u.Name)
+			*vmss.Sku.Capacity = u.DesiredReplicas + u.MaxSurge
```

Comment:
I will change u.MaxSurge to 1 for now and I'll open an issue about extending it in the future.